text/0003-extended-error-management.md

```diff
@@ -0,0 +1,117 @@
+
+For context, here is an example of a handler: https://github.com/jules2689/extended_bundler-errors/blob/master/lib/extended_bundler/handlers/gda.yml
+
+Troubleshoot happens like this:
+
+```ruby
+def troubleshoot(spec_install)
+  # Determine a path in the plugin gem to a "handlers" folder, including the name of the gem
+  # E.g. ~/path/to/gem/lib/handlers/nokogiri.yml
+  path = handler_path(spec_install.name)
+  return nil unless File.exist?(path)
+
+  # If we have a handler, then load it
+  # Each handler may contain multiple matchable errors, iterate and match
+  # Each matchable error is based on the gem version and a pattern to match
+  # If we match the gem version and the pattern is found in the original error message, then we extract out the build error and replace it. 
+  troubleshooted = false
+  yaml = YAML.load_file(path)
+  yaml.each do |handler|
+    next unless version_match?(spec_install.spec.version, handler['versions'])
+    next unless handler['matching'].any? { |m| spec_install.error =~ Regexp.new(m) }
+    spec_install.error = build_error(spec_install, handler)
+    troubleshooted = true
+  end
+
+  # If we did not troubleshoot, and this was a native extension issue
+  # then tell the user to file an issue to help us track down more bugs
+  if !troubleshooted && spec_install.error.include?('Failed to build gem native extension')
+    spec_install.error = spec_install.error + "\n\n" + NATIVE_EXTENSION_MSG
+  end
+end
+```
+
+In Bundler, we would largely do the same.
+
+- We would have a collection of handlers which match on name, version, and pattern in the error.
+- Instead of a plugin hook, we would just call at the same point that the plugin hook is called.
+
+Potential improvements
+---
+- A new plugin hook could be added that allows plugins to register their own handlers
+- Gemspecs could specify error handlers for itself. This would allow gems to debug their own installation issues.
+
+# Drawbacks
+
+1. We would need to field issues that include GCC dumps. We would then need to make handlers and include them in Bundler. If you don't update bundler, then you don't get new errors (unless we include the plugin hook and make errors updateable that way, or make some remote index).
+2. It might be better to go deeper into the Ruby ecosystem and ultimately include this in Rubygems itself
+3. It may obfuscate error messages a bit more and people won't ever deal with them, making the knowledge of GCC dumps even more niche
+
+# Rationale and Alternatives
+
+- This is a fairly simple integration point that is fairly easily removed if needed
+- This maintains all current information (in the file) and provides quick and easy understanding for GCC errors by developers
+
+### Alternatives
+
+- An alternative would be to drastically improve the plugin system, but I suspect that will take a lot of time and this issue affects everyone - so why not make it core?
```

Comment:
What specifically is missing in the plugin system that gets in the way of this plugin being as good as possible?

Reply:
I still intend to write this up, but it's mainly adoption. People aren't going to install bundler, then install a bunch of gems. It's just not a thing and will take a long time to make that standard. With all the tutorials and lessons on installing ruby/bundler, and tribal knowledge in the world, no one is gonna say "oh yea, need to bundle plugin install blah".

There's also technical issues which I'll get into in the doc I write up.